docs/cip/CIP-07102024_Write_Ahead_Log_Pruning_Vacuuming.md

```diff
@@ -0,0 +1,331 @@
+# CIP-07102024: Write-Ahead Log Pruning & Vacuuming
+
+## Status
+
+Current Status: `Under Discussion`
+
+## Motivation
+
+Chroma's SQLite-based write-ahead log grows infinitely over time. When ingesting large amounts of data, it's not uncommon for the SQLite database to grow to many gigabytes in size. Large databases cost more, take longer to back up, and can result in decreased query performance.
+
+There are two separate problems:
+
+- The database, specifically the `embeddings_queue` table, has unbounded growth.
+- The SQLite `VACUUM` command, often recommended for such scenarios, takes an exclusive lock on the database and is potentially quite slow.
+
+This CIP addresses both issues.
+
+## Proposed Changes
+
+After every write transaction, if log pruning is enabled, the `embeddings_queue` table will be pruned to remove rows that are no longer needed. Specifically, rows with a sequence ID less than the minimum sequence ID of any active subscriber will be deleted. (As long as this is done continuously, this is a relatively cheap operation.)
+
+This does not directly reduce the disk size of the database, but allows SQLite to reuse the space occupied by the deleted rows—thus effectively bounding the disk usage of the `embeddings_queue` table by `hnsw:sync_threshold`.
+
+To control log pruning, `SqlEmbeddingsQueue` will get a new configuration object with a single parameter: `automatically_prune`. This will default to `false` for systems with non-empty embedding queues, because:
+
+- The first pruning operation for a large embeddings queue can be very slow.
+- Some users may be relying on the WAL as a full backup.
+
+If the system's embedding queue is empty (a fresh system), `automatically_prune` will default to `true`.
+
+This configuration object will be stored in a new table, `embeddings_queue_config`.
```

Comment:
after playing around with it I believe it makes more sense to store this as a property of the embeddings queue rather than as a property of the system
e.g. this setting is meaningless when not using the SQLite implementation of the embeddings queue (aka distributed), so at the system level the setting is only meaningful in some contexts

Reply:
draft implementation of what this looks like here: https://github.com/chroma-core/chroma/pull/2557

Reply:
This was my only question. I'm happy with this implementation.